Added blacklisted_paths option

schroedingerskatze42 opened the pull request:
- adds a list of strings (regular expressions) to describe patterns of paths not to be cached.

| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    |yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | -
| Documentation   | -
| License         | MIT


#### What's in this PR?

Added an option 'blacklisted_paths'; a list of regular expressions to describe paths not to be cached. 


#### Why?

This is necessary if a third party resource is requested of providers, who don't get the usage of response cache headers. (But will return no-cache every time, for every endpoint)  


#### Example Usage

``` php
$cachePlugin = new CachePlugin($pool, $streamFactory, ['blacklisted_paths' => [
  '^\/this-path-shall\/not-been-cached\/.*',
]]);
```


#### Checklist

- [X] Updated CHANGELOG.md to describe BC breaks / deprecations | new feature | bugfix
- [X] Documentation pull request created: https://github.com/php-http/documentation/pull/274

Commit:
Added blacklisted_paths option
 - adds a list of strings (regular expressions) to describe patterns of paths not to be cached.

Commit:
Fix CI

Commit:
Fix CI # 2
 - honestly, the lengths of a line should be fixed.

Commit:
Fix CI # 3
 - that has been a fine restriction

dbu commented on src/CachePlugin.php:
```diff
@@ -72,10 +73,7 @@ public function __construct(CacheItemPoolInterface $pool, StreamFactory $streamF
-                'You can\'t provide config option "respect_cache_headers" and "respect_response_cache_directives". '.
-                'Use "respect_response_cache_directives" instead.'
-            );
+            throw new \InvalidArgumentException('You can\'t provide config option "respect_cache_headers" and "respect_response_cache_directives". '.'Use "respect_response_cache_directives" instead.');
```
did the codestyle checker complain about this before? otherwise lets not touch it.

if codestyle complains, lets remove the concatenation and make it all one string.

schroedingerskatze42 replied:
yes, the code style checker complained about it; should already have been existed before. no magic button (auto format) used.

regarding the test: WIP, give me 15 minutes.

dbu replied:
the stupid thing is that it does a string concat on that one line now, for no good reason... can you remove the `'.'` in the middle please?

schroedingerskatze42 replied:
done.

dbu commented on src/CachePlugin.php:
```diff
@@ -353,13 +358,15 @@ private function configureOptions(OptionsResolver $resolver)
             'respect_response_cache_directives' => ['no-cache', 'private', 'max-age', 'no-store'],
             'cache_key_generator' => null,
             'cache_listeners' => [],
+            'blacklisted_paths' => [],  // restricted for
```
please remove the comment, the phpdoc on the constructor should document this (plus the documentation change)

schroedingerskatze42 replied:
done

dbu commented on src/CachePlugin.php:
```diff
@@ -246,16 +244,23 @@ private function calculateResponseExpiresAt($maxAge)
      * @return bool
      */
-    protected function isCacheable(ResponseInterface $response)
+    protected function isCacheable(RequestInterface $request, ResponseInterface $response)
```
i wonder why this method is protected. extending the plugin sounds like a bad idea, but maybe that was the intention. as it is protected, this is technically a BC break. we could separate this into another method `isCacheableRequest` to avoid the BC break.

schroedingerskatze42 replied:
you are right. i outsourced the checking of the request into the method isCacheableRequest and reset the "isCacheable" to it's original state.

The only thing, I am bothering about is the naming (one of the two hard things of computer science). If we add a method "isCacheableRequest", the method "isCacheable" should be renamed to "isCacheableResponse" and a new method "isCacheable" should take both arguments and return the result of isCacheableRequest && isCachableResponse().

This however would lead again to the BC, we would like to avoid.

dbu replied:
yep. my idea would be that in version 2 we make the plugin final, this method private and can change the signature.

dbu commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
 # Change Log
 
+## 1.6.1 - 2019-11-25
```
this is a new feature and should lead to 1.7.0

please replace the date with "unreleased"

schroedingerskatze42 replied:
done; also replaced 1.6.1 with 1.7.0; not quite sure, what is right here. feel free to enlighten me :-)

dbu replied:
that looks good, thanks.

dbu reviewed (commented):
thanks for contributing. this is rather an edge case, but as we do have modes of ignoring caching instructions from the server, it seems reasonable to me to have this config.

could you please add something to the test in spec/CachePluginSpec.php to have one example of blacklisting tested? and wdyt of splitting it into a separate method to avoid the BC break?

Commit:
Fixing Pull Request Requests
 - outsourced checking of blacklisted_paths to new method "isCacheableRequest"
 - removed related functionality from "isCacheable"
 - removed obsolete comment

dbu commented on src/CachePlugin.php:
```diff
@@ -266,6 +264,24 @@ protected function isCacheable(ResponseInterface $response)
+     *
+     * @return bool
+     */
+    protected function isCacheableRequest(RequestInterface $request)
```
please make this method private.

schroedingerskatze42 replied:
done.

dbu commented:
apart from my nitpicks, i am happy with this.

@nobYsDarling can you please do the whitespace fix that codestyle is complaining about?

@Nyholm any inputs from your side or happy with this?

Commit:
Fixing Pull Request Requests
 - added spec tests for blacklisted paths
  - it_does_not_store_responses_of_requests_to_blacklisted_paths
  - it_stores_responses_of_requests_not_in_blacklisted_paths

Commit:
Fixing CI

Commit:
Implemented Feedback
 - removed unnecessary string concatenation
 - set isCacheableRequest to private
